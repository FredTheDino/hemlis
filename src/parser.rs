#[cfg(test)]
mod tests {
    use insta::assert_snapshot;

    macro_rules! gen_parser {
        ($a:ident, $p:ident) => {
            pub fn $a<'t>(src: &'t str) -> String {
                pub fn inner<'t>(src: &'t str) -> Result<String, String> {
                    use crate::{ast::Ast, grammer, lexer};
                    use std::io::BufWriter;

                    let parser = grammer::$p::new();
                    let out = parser
                        .parse(
                            lexer::lex(&src)
                                .into_iter()
                                .map(|(token, span)| Ok((span.start, token?, span.end))),
                        )
                        .map_err(|x| format!("{:?}", x))?;

                    let mut buf = BufWriter::new(Vec::new());
                    out.show(0, &mut buf).map_err(|x| format!("{:?}", x))?;
                    let inner = buf.into_inner().map_err(|x| format!("{:?}", x))?;
                    String::from_utf8(inner).map_err(|x| format!("{:?}", x))
                }
                match inner(src) {
                    Ok(s) | Err(s) => s,
                }
            }
        };
    }

    gen_parser!(p_header, HeaderParser);
    gen_parser!(p_import, ImportParser);

    #[test]
    fn empty_string() {
        assert_snapshot!(p_header(""));
    }

    #[test]
    fn normal_definition() {
        assert_snapshot!(p_header("module A (a, b, c) where"));
    }

    #[test]
    fn everything_header() {
        assert_snapshot!(p_header(r#"
module A.B.C (a, class B, C, D(..), E(F, G), 
 (+), type (+), module H) where

import A as A
import A.B.C as A.C
import A.B.C (a, class B, 
 C, D(..), E(F, G), (+), type (+))
import A.B.C hiding (foo)
        "#));
    }

    #[test]
    fn simple_import() {
        assert_snapshot!(p_import("import A as A"))
    }

    #[test]
    fn import_with_newlines() {
        assert_snapshot!(p_import("import A (foo\n , bar\n , baz)"))
    }
}
