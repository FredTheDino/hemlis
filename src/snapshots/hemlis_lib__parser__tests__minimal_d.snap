---
source: src/parser.rs
expression: "p_module(r\"\nmodule D where\n\ntype D = X.Z (A (B.C D)) E\n            \")"
---
19 of 19
===
Module
 Header
  MName
   (1, 7)->(1, 8)
  NULL
  (1, 0)->(1, 6)
  (1, 9)->(1, 14)
  Decl::Type
   ProperName
    (3, 5)->(3, 6)
   Typ::App
    Typ::App
     Typ::Constructor
      QProperName
       Qual
        (3, 9)->(3, 10)
       ProperName
        (3, 11)->(3, 12)
     Typ::Paren
      Typ::App
       Typ::Constructor
        QProperName
         NULL
         ProperName
          (3, 14)->(3, 15)
       Typ::Paren
        Typ::App
         Typ::Constructor
          QProperName
           Qual
            (3, 17)->(3, 18)
           ProperName
            (3, 19)->(3, 20)
         Typ::Constructor
          QProperName
           NULL
           ProperName
            (3, 21)->(3, 22)
    Typ::Constructor
     QProperName
      NULL
      ProperName
       (3, 25)->(3, 26)
===
